src/ziggurat/messaging/rabbitmq/cluster/producer.clj

```diff
@@ -0,0 +1,83 @@
+(ns ziggurat.messaging.rabbitmq.cluster.producer
+  (:require [clojure.tools.logging :as log]
+            [langohr.channel :as lch]
+            [langohr.basic :as lb]
+            [langohr.http :as lh]
+            [taoensso.nippy :as nippy]
+            [langohr.exchange :as le]
+            [ziggurat.messaging.rabbitmq.retry :refer :all]
+            [langohr.queue :as lq]
+            [clojure.string :as str]
+            [clojure.set :as set])
+  (:import (org.apache.kafka.common.header.internals RecordHeader)))
+
+(defn- record-headers->map [record-headers]
+  (reduce (fn [header-map record-header]
+            (assoc header-map (.key record-header) (String. (.value record-header))))
+          {}
+          record-headers))
+
+(defn- properties-for-publish
+  [expiration headers]
+  (let [props {:content-type "application/octet-stream"
+               :persistent   true
+               :headers      (record-headers->map headers)}]
+    (if expiration
+      (assoc props :expiration (str expiration))
+      props)))
+
+(defn publish
+  ([connection exchange message-payload expiration]
+   (try
+     (with-retry {:count      5
+                  :wait       100
+                  :on-failure #(log/error "publishing message to rabbitmq failed with error " (.getMessage %))}
+       (with-open [ch (lch/open connection)]
+         (lb/publish ch exchange "" (nippy/freeze (dissoc message-payload :headers))
+                     (properties-for-publish expiration (:headers message-payload)))))
+     (catch Throwable e
+       (log/error e "Pushing message to rabbitmq failed, data: " message-payload)
+       (throw (ex-info "Pushing message to rabbitMQ failed after retries, data: " {:type  :rabbitmq-publish-failure
+                                                                                   :error e}))))))
+
+(defn set-ha-policy [queue-name cluster-config]
+  (let [hosts (atom (str/split (:hosts cluster-config) #","))]
+    (with-retry {:count (count @hosts)
+                 :wait 50
+                 :on-failure #(log/error "setting ha-policies failed " (.getMessage %))}
+      (binding [lh/*endpoint* (str "http://" (ffirst (swap-vals! hosts rest)) ":" (get cluster-config :admin-port 15672))
+                lh/*username* (:username cluster-config)
+                lh/*password* (:password cluster-config)]
+        (log/info "Applying HA Policies to queue: " queue-name)
+        (lh/set-policy "/" (str queue-name "_ha_policy")
```

Comment:
here, we are only taking the queue-name, where are the ha policies being set for exchanges?

Reply:
fixed it